spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/core/SpannerTemplateTests.java

```diff
@@ -287,49 +284,48 @@ public void readOnlyTransactionTest() {
     verify(readOnlyTransaction, times(2)).read(eq("custom_test_table"), any(), any());
   }
 
-  @Test
-  public void readOnlyTransactionDmlTest() {
 
-    this.expectedException.expectMessage("A read-only transaction template cannot execute DML.");
+  @Test
+  <T> void readOnlyTransactionDmlTest() {
```

Comment:
Generics are useful when the callers of this method may need to provide different types. Since this method is only called by the JUnit framework, and the test itself is defining the operation lambda, it is at this point known what the return type is. Well, kind of. The lambda returns `null`, so the type could be anything -- Long (because that would have been the result of `executeDmlStatement()`), Object (technically correct for any return value. In fact, the original code had Java infer Object as the type), or, my personal favorite when return value is `null` and return type does not matter, `Void`.

Looking at javadoc for `performReadOnlyTransaction`, it says "@param <T> the final return type of the operations." Since the return value of all these test lambdas is `null`, I'd use `Void` for the second parameterized type.